pkg/bpf/map_linux_test.go

```diff
@@ -567,6 +568,57 @@ func (s *BPFPrivilegedTestSuite) TestCheckAndUpgrade(c *C) {
 					path, _ := m.Path()
 					os.Remove(path)
 
+					m.Close()
+				}
+			},
+		},
+		{
+			name: "MapTypeLRUHash on 4.9 kernel: no prealloc to prealloc upgrade",
+			run: func() []*Map {
+				// Asserts that maps with type MapTypeLRUHash on 4.9 kernels
+				// where preallocation is disabled can be recreated without
+				// requiring them to be removed due to a flag mismatch.
+
+				// Specify 4.9 kernel supported maps types and disable preallocation.
+				SetMapTypesFromProber(newProber(mapTypes49))
+				DisableMapPreAllocation()
+
+				upgradeMap := NewMap("cilium_test_upgrade",
+					MapTypeLRUHash,
+					&TestKey{},
+					int(unsafe.Sizeof(TestKey{})),
+					&TestValue{},
+					int(unsafe.Sizeof(TestValue{})),
+					maxEntries,
+					0,
+					0,
+					ConvertKeyValue).WithCache()
+				_, err := upgradeMap.OpenOrCreate()
+				c.Assert(err, IsNil)
+
+				upgrade := upgradeMap.CheckAndUpgrade(&upgradeMap.MapInfo)
+				c.Assert(upgrade, Equals, false)
```

Comment:
If I follow right, the `upgrade == false` case here is basically saying that:
* Typically, LRU requires pre-allocation.
* Given the underlying lack of LRU support from the "kernel" prober, this map type would actually use hashmap
* Since the map type is switched to hashmap, now pre-allocation *can* be disabled
* When we try to upgrade the map, defining that its type should be LRU, there's no intermediate state where we decide that the map *should* be upgraded because the desired type is LRU (or the preallocation flags are mismatched)
* Instead, every single time the map info is evaluated, the type & flags are evaluated first and then the upgrade decision is made based on the attributes afterwards.

At a glance I found this a little bit cryptic in the reading of the test, you might consider stashing a comment something like this in the code. But not such a big deal.

Reply:
Thanks, I'll add this snippet as a comment. It can indeed be hard to follow at first glance, as it's very subtle. Also amended the commit to include a reference to the commit that fixed this to establish a link between this regression test and the fix.